File: electron/src-tauri/src/core/prompt.rs
use crate::core::tools::Tool;

pub trait PromptPart {
    fn render(&self) -> String;
}

pub struct IdentityPart {
    pub user_name: String,
}

impl PromptPart for IdentityPart {
    fn render(&self) -> String {
        format!(
            "You are MOSAIC, a highly capable AI agent operating in a terminal-like environment.\n\
             Your goal is to assist the user, {}, by executing tools and providing information.",
            self.user_name
        )
    }
}

pub struct ToolFormatPart<'a> {
    pub tools: &'a [Box<dyn Tool>],
}

impl<'a> PromptPart for ToolFormatPart<'a> {
    fn render(&self) -> String {
        let tools_json: Vec<serde_json::Value> = self.tools.iter().map(|t| {
            serde_json::json!({
                "name": t.name(),
                "description": t.description(),
                "parameters": serde_json::from_str::<serde_json::Value>(&t.parameters()).unwrap_or(serde_json::json!({}))
            })
        }).collect();

        format!(
            "AVAILABLE TOOLS:\n{}\n\n\
             TOOL CALLING FORMAT:\n\
             To call a tool, use the following XML-like format. All parameter values MUST be strings:\n\
             <tool_call>\n  <name>tool_name</name>\n  <parameters>\n    <your_parameter_name>the_actual_value</your_parameter_name>\n  </parameters>\n</tool_call>\n\n\
             EXAMPLE:\n\
             <tool_call>\n  <name>read_file</name>\n  <parameters>\n    <path>README.md</path>\n  </parameters>\n</tool_call>\n\n\
             You can call only one tool at a time. After a tool call, the system will provide the result in a <tool_result> block.\n\
             Wait for the result before proceeding.",
            serde_json::to_string_pretty(&tools_json).unwrap_or_default()
        )
    }
}

pub struct ChecklistBehaviorPart;

impl PromptPart for ChecklistBehaviorPart {
    fn render(&self) -> String {
        "MODIFIED BEHAVIOR: CHECKLISTS\n\
         You MUST maintain a checklist of your progress using the 'manage_todos' tool.\n\
         1. When you start a complex task, initialize the checklist.\n\
         2. For every step you take, update the checklist status.\n\
         3. Use the following format for each item in the 'checklist' parameter:\n\
            - [ ] for pending tasks\n\
            - [>] for in-progress tasks\n\
            - [x] for completed tasks\n\
            - Follow with 'Task Name <- context/details'\n\
         EXAMPLE CALL:\n\
         <tool_call>\n  <name>manage_todos</name>\n  <parameters>\n    <checklist>\n[x] Initialization <- Done\n[>] Researching files <- Currently reading routes\n[ ] Final report <- Pending\n    </checklist>\n</tool_call>\n\
         Always inform the user about the current state of the checklist.".to_string()
    }
}

pub struct PersonaPart {
    pub persona: String,
}

impl PromptPart for PersonaPart {
    fn render(&self) -> String {
        format!("## YOUR PERSONA\n\n{}", self.persona)
    }
}

pub struct PromptBuilder;

impl PromptBuilder {
    pub fn create_system_prompt(tools: &[Box<dyn Tool>], _workspace: &str, user_name: &str, persona: Option<String>) -> String {
        let mut parts: Vec<Box<dyn PromptPart>> = Vec::new();
        
        if let Some(p) = persona {
            parts.push(Box::new(PersonaPart { persona: p }));
        }
        
        parts.push(Box::new(IdentityPart { user_name: user_name.to_string() }));
        parts.push(Box::new(ToolFormatPart { tools }));
        parts.push(Box::new(ChecklistBehaviorPart));

        let rendered: Vec<String> = parts.iter().map(|p| p.render()).collect();
        
        rendered.join("\n\n") + "\n\n\
        ## CRITICAL RULES\n\n\
        1. **You MUST either call a tool OR provide a final answer. Never say what you're \"going to do\" - just DO IT.**\n\
        2. If you need more information, call the appropriate tool immediately.\n\
        3. Only provide a final answer when you have completed the task and gathered all necessary information.\n\
        4. Never respond with \"Now let me...\" or \"I will...\" - if you need to do something, call the tool."
    }

    pub fn format_tool_result(name: &str, result: &str) -> String {
        format!("<tool_result name=\"{}\">\n{}\n</tool_result>", name, result)
    }
}

File: vscode/src/core/prompt.ts
import { TOOL_CALL_START, TOOL_CALL_END, TOOL_RESPONSE_START, TOOL_RESPONSE_END } from "./protocol";

export interface ToolDescription {
  name: string;
  description: string;
  schema: any;
}

export class PromptBuilder {
  static createSystemPrompt(tools: ToolDescription[], workspaceName: string, userName: string): string {
    const toolsJson = tools.map(t => ({
      name: t.name,
      description: t.description,
      parameters: t.schema
    }));

    const hasTodoTools = tools.some(t => t.name.includes('todo'));

    const todoInstructions = hasTodoTools ? `
2. **Task Management (CRITICAL)**: You MUST maintain a list of active tasks using the todo tools. Creating and updating TODOs is REALLY IMPORTANT for the task you need to do. It allows the user to follow your progress. Start by listing/creating your plan in the TODO list.
` : `
2. **Execution Phase**: Proceed directly to executing the requested actions.
`;

    return `
You are Mosaic, a powerful AI assistant integrated into VSCode, helping ${userName} in the workspace: ${workspaceName}.
You are a function calling AI model. You are provided with function signatures within <tools> </tools> XML tags. 
You may call one or more functions to assist with the user query. 
If available tools are not relevant in assisting with user query, just respond in natural conversational language. 
Don't make assumptions about what values to plug into functions. 

# TOOLS
<tools>
${JSON.stringify(toolsJson, null, 2)}
</tools>

# TOOL CALL FORMAT
For each function call return a JSON object, with the following pydantic model json schema for each:
{'title': 'FunctionCall', 'type': 'object', 'properties': {'name': {'title': 'Name', 'type': 'string'}, 'arguments': {'title': 'Arguments', 'type': 'object'}}, 'required': ['name', 'arguments']}

Each function call should be enclosed within ${TOOL_CALL_START} ${TOOL_CALL_END} XML tags.
IMPORTANT: Do NOT use markdown code blocks (like \`\`\`json) inside the tags. Just the raw JSON object.
Example:
${TOOL_CALL_START}
{"name": "tool_name", "arguments": {"param1": "value1"}}
${TOOL_CALL_END}

# TOOL RESPONSE FORMAT
After calling & executing the functions, you will be provided with function results within ${TOOL_RESPONSE_START} ${TOOL_RESPONSE_END} XML tags. The system will automatically link responses to calls using unique IDs.
Example:
${TOOL_RESPONSE_START.replace('>', ' id="call_123">')}
{"tool_call_id": "call_123", "name": "tool_name", "content": "result"}
${TOOL_RESPONSE_END}

# CODING WORKFLOW
1. **Thinking Phase**: Always start your response with a thinking phase wrapped in <thought> tags. Provide a brief internal monologue explaining your reasoning and what you've found so far.${todoInstructions}
   - **Tip**: Use \`clear_todos\` at the start of a new major request if the existing list is irrelevant.
   - **Tip**: Always \`list_todos\` or rely on the tool output after an update to keep track of IDs.
3. **Action Phase**: Call exactly one tool or provide a final answer based on your thinking.
4. **Final Answer**: When you have completed the requested task or reached a stopping point, you MUST provide a concise summary (resume) of all the actions you took and the results achieved.

# CRITICAL RULES
- **Workspace Structure**: Explore the workspace using \`list_directory\` and \`read_file\` to understand the project architecture.
- **Relative Paths**: All your actions and tool calls MUST use relative paths from the workspace root. You do not need to know the absolute path of the workspace on the host machine.
- **Directories vs Files**: \`read_file\` ONLY works on files. Use \`list_directory\` for folders.
- **Always Resume**: Never end a conversation without explaining what was done.
- **Task Completion**: When you have fully completed the user's request and no further actions are needed, you MUST include the tag \`<task_finished />\` at the very end of your message. This signals to the system that the agentic loop can stop.
- **Style**: Maintain a professional tone and avoid excessive emojis.
- **Preserve History**: NEVER delete the \`.git\` or \`.mosaic\` directories.
`;
  }

  static formatToolResult(name: string, result: any, callId: string): string {
    const content = typeof result === 'string' ? result : JSON.stringify(result, null, 2);
    const responseJson = JSON.stringify({
      tool_call_id: callId,
      name: name,
      content: content
    }, null, 2);
    return `${TOOL_RESPONSE_START.replace('>', ` id="${callId}">`)}\n${responseJson}\n${TOOL_RESPONSE_END}`;
  }
}
